src/Evolution/Systems/RelativisticEuler/Valencia/PrimitiveFromConservative.cpp

```diff
@@ -0,0 +1,181 @@
+                const auto& equation_of_state) noexcept {
+              return equation_of_state->pressure_from_density(
+                  Scalar<double>(rho));
+            },
+            [&rho, &epsilon ](
+                const std::integral_constant<size_t, 2>& /*thermodynamic_dim*/,
+                const auto& equation_of_state) noexcept {
+              return equation_of_state->pressure_from_density(
+                  Scalar<double>(rho), Scalar<double>(epsilon));
+            })(std::integral_constant<size_t, ThermodynamicDim>{},
+               equation_of_state_)
+            .get();
+
+    const double a = pressure / e;
+    const double h = (1.0 + epsilon) * (1.0 + a);
+    return z - r / h;
+  }
+
+ private:
+  const Scalar<DataType>& tilde_d_;
+  const Scalar<DataType>& tilde_tau_;
+  const Scalar<DataType>& tilde_s_magnitude_;
+  const std::unique_ptr<
+      EquationsOfState::EquationOfState<true, ThermodynamicDim>>&
+      equation_of_state_;
+};
+}  // namespace
+
+template <typename DataType, size_t Dim, size_t ThermodynamicDim>
+void primitive_from_conservative(
+    const gsl::not_null<Scalar<DataType>*> rest_mass_density,
+    const gsl::not_null<Scalar<DataType>*> specific_internal_energy,
+    const gsl::not_null<Scalar<DataType>*> lorentz_factor,
+    const gsl::not_null<Scalar<DataType>*> specific_enthalpy,
+    const gsl::not_null<Scalar<DataType>*> pressure,
+    const gsl::not_null<tnsr::I<DataType, Dim, Frame::Inertial>*>
+        spatial_velocity,
+    const Scalar<DataType>& tilde_d, const Scalar<DataType>& tilde_tau,
+    const tnsr::i<DataType, Dim, Frame::Inertial>& tilde_s,
+    const tnsr::II<DataType, Dim, Frame::Inertial>& inv_spatial_metric,
+    const Scalar<DataType>& sqrt_det_spatial_metric,
+    const std::unique_ptr<
+        EquationsOfState::EquationOfState<true, ThermodynamicDim>>&
+        equation_of_state) noexcept {
+  const auto tilde_s_M = raise_or_lower_index(tilde_s, inv_spatial_metric);
+  const Scalar<DataType> tilde_s_magnitude{
+      sqrt(get(dot_product(tilde_s, tilde_s_M)))};
+
+  // find z via root find
+  const DataType k = get(tilde_s_magnitude) / (get(tilde_tau) + get(tilde_d));
+  const DataType lower_bound = 0.5 * k / sqrt(1.0 - 0.25 * square(k));
+  const DataType upper_bound = k / sqrt(1.0 - square(k));
+  const auto f_of_z = FunctionOfZ<DataType, ThermodynamicDim>{
+      tilde_d, tilde_tau, tilde_s_magnitude, equation_of_state};
+
+  const auto z =
+      // NOLINTNEXTLINE
+      RootFinder::toms748(f_of_z, lower_bound, upper_bound,
+                          10.0 * std::numeric_limits<double>::epsilon(),
+                          10 * std::numeric_limits<double>::epsilon(), 100);
+  get(*lorentz_factor) = sqrt(1.0 + square(z));
+  get(*rest_mass_density) =
+      get(tilde_d) / get(*lorentz_factor) / get(sqrt_det_spatial_metric);
+  get(*specific_internal_energy) =
+      (get(*lorentz_factor) * get(tilde_tau) - z * get(tilde_s_magnitude)) /
```

Comment:
I think this tilde_tau also needs to be divided by tilde_d, as in the next term in the eqn. Can you check?

Reply:
@teukolsky that is true, but I am dividing both terms...